From 2bc28a61a1c8bdd439f812cdb9908bf8097c6f0b Mon Sep 17 00:00:00 2001
From: Haris Okanovic <haris.okanovic@ni.com>
Date: Mon, 21 Jul 2014 09:23:45 -0500
Subject: [PATCH] Decorate API functions on Windows with __cdecl

Adds __cdecl decoration to API functions on Win32 systems so that cURL
may be used in applications compiled with /Gz (__stdcall).

Upstream-Status: Pending [Needs clean-up]
---
 include/curl/curl.h  | 106 ++++++++++++++++++++++---------------------
 include/curl/easy.h  |  16 +++----
 include/curl/multi.h |  32 ++++++-------
 3 files changed, 79 insertions(+), 75 deletions(-)

diff --git a/include/curl/curl.h b/include/curl/curl.h
index b7cb30a..c9c9b7b 100644
--- a/include/curl/curl.h
+++ b/include/curl/curl.h
@@ -112,6 +112,12 @@ typedef void CURLSH;
  * libcurl external API function linkage decorations.
  */
 
+#if defined(WIN32) || defined(_WIN32)
+#  define CURL_FUNCC __cdecl
+#else
+#  define CURL_FUNCC
+#endif /* defined(WIN32) || defined(_WIN32) */
+
 #ifdef CURL_STATICLIB
 #  define CURL_EXTERN
 #elif defined(CURL_WIN32) || defined(__SYMBIAN32__) || \
@@ -217,7 +223,7 @@ struct curl_httppost {
 
 /* This is the CURLOPT_PROGRESSFUNCTION callback prototype. It is now
    considered deprecated but was the only choice up until 7.31.0 */
-typedef int (*curl_progress_callback)(void *clientp,
+typedef int (CURL_FUNCC * curl_progress_callback)(void *clientp,
                                       double dltotal,
                                       double dlnow,
                                       double ultotal,
@@ -226,7 +232,7 @@ typedef int (*curl_progress_callback)(void *clientp,
 /* This is the CURLOPT_XFERINFOFUNCTION callback prototype. It was introduced
    in 7.32.0, avoids the use of floating point numbers and provides more
    detailed information. */
-typedef int (*curl_xferinfo_callback)(void *clientp,
+typedef int (CURL_FUNCC *curl_xferinfo_callback)(void *clientp,
                                       curl_off_t dltotal,
                                       curl_off_t dlnow,
                                       curl_off_t ultotal,
@@ -257,8 +263,7 @@ typedef int (*curl_xferinfo_callback)(void *clientp,
 /* This is a magic return code for the write callback that, when returned,
    will signal libcurl to pause receiving on the current transfer. */
 #define CURL_WRITEFUNC_PAUSE 0x10000001
-
-typedef size_t (*curl_write_callback)(char *buffer,
+typedef size_t (CURL_FUNCC * curl_write_callback)(char *buffer,
                                       size_t size,
                                       size_t nitems,
                                       void *outstream);
@@ -326,7 +331,7 @@ struct curl_fileinfo {
 /* if splitting of data transfer is enabled, this callback is called before
    download of an individual chunk started. Note that parameter "remains" works
    only for FTP wildcard downloading (for now), otherwise is not used */
-typedef long (*curl_chunk_bgn_callback)(const void *transfer_info,
+typedef long (CURL_FUNCC *curl_chunk_bgn_callback)(const void *transfer_info,
                                         void *ptr,
                                         int remains);
 
@@ -340,7 +345,7 @@ typedef long (*curl_chunk_bgn_callback)(const void *transfer_info,
    Even if downloading of this chunk was skipped in CHUNK_BGN_FUNC.
    This is the reason why we don't need "transfer_info" parameter in this
    callback and we are not interested in "remains" parameter too. */
-typedef long (*curl_chunk_end_callback)(void *ptr);
+typedef long (CURL_FUNCC *curl_chunk_end_callback)(void *ptr);
 
 /* return codes for FNMATCHFUNCTION */
 #define CURL_FNMATCHFUNC_MATCH    0 /* string corresponds to the pattern */
@@ -349,7 +354,7 @@ typedef long (*curl_chunk_end_callback)(void *ptr);
 
 /* callback type for wildcard downloading pattern matching. If the
    string matches the pattern, return CURL_FNMATCHFUNC_MATCH value, etc. */
-typedef int (*curl_fnmatch_callback)(void *ptr,
+typedef int (CURL_FUNCC *curl_fnmatch_callback)(void *ptr,
                                      const char *pattern,
                                      const char *string);
 
@@ -358,7 +363,7 @@ typedef int (*curl_fnmatch_callback)(void *ptr,
 #define CURL_SEEKFUNC_FAIL     1 /* fail the entire transfer */
 #define CURL_SEEKFUNC_CANTSEEK 2 /* tell libcurl seeking can't be done, so
                                     libcurl might try other means instead */
-typedef int (*curl_seek_callback)(void *instream,
+typedef int (CURL_FUNCC *curl_seek_callback)(void *instream,
                                   curl_off_t offset,
                                   int origin); /* 'whence' */
 
@@ -376,7 +381,7 @@ typedef int (*curl_seek_callback)(void *instream,
   want to abort the request */
 #define CURL_TRAILERFUNC_ABORT 1
 
-typedef size_t (*curl_read_callback)(char *buffer,
+typedef size_t (CURL_FUNCC * curl_read_callback)(char *buffer,
                                       size_t size,
                                       size_t nitems,
                                       void *instream);
@@ -397,7 +402,7 @@ typedef enum {
                                 CURLE_ABORTED_BY_CALLBACK */
 #define CURL_SOCKOPT_ALREADY_CONNECTED 2
 
-typedef int (*curl_sockopt_callback)(void *clientp,
+typedef int (CURL_FUNCC *curl_sockopt_callback)(void *clientp,
                                      curl_socket_t curlfd,
                                      curlsocktype purpose);
 
@@ -412,12 +417,12 @@ struct curl_sockaddr {
 };
 
 typedef curl_socket_t
-(*curl_opensocket_callback)(void *clientp,
+(CURL_FUNCC * curl_opensocket_callback)(void *clientp,
                             curlsocktype purpose,
                             struct curl_sockaddr *address);
 
 typedef int
-(*curl_closesocket_callback)(void *clientp, curl_socket_t item);
+(CURL_FUNCC *curl_closesocket_callback)(void *clientp, curl_socket_t item);
 
 typedef enum {
   CURLIOE_OK,            /* I/O operation successful */
@@ -432,7 +437,7 @@ typedef enum {
   CURLIOCMD_LAST         /* never use */
 } curliocmd;
 
-typedef curlioerr (*curl_ioctl_callback)(CURL *handle,
+typedef curlioerr (CURL_FUNCC * curl_ioctl_callback)(CURL *handle,
                                          int cmd,
                                          void *clientp);
 
@@ -443,11 +448,11 @@ typedef curlioerr (*curl_ioctl_callback)(CURL *handle,
  * curl_global_init_mem() function to set user defined memory management
  * callback routines.
  */
-typedef void *(*curl_malloc_callback)(size_t size);
-typedef void (*curl_free_callback)(void *ptr);
-typedef void *(*curl_realloc_callback)(void *ptr, size_t size);
-typedef char *(*curl_strdup_callback)(const char *str);
-typedef void *(*curl_calloc_callback)(size_t nmemb, size_t size);
+typedef void *(CURL_FUNCC * curl_malloc_callback)(size_t size);
+typedef void (CURL_FUNCC * curl_free_callback)(void *ptr);
+typedef void *(CURL_FUNCC * curl_realloc_callback)(void *ptr, size_t size);
+typedef char *(CURL_FUNCC * curl_strdup_callback)(const char *str);
+typedef void *(CURL_FUNCC * curl_calloc_callback)(size_t nmemb, size_t size);
 
 #define CURL_DID_MEMORY_FUNC_TYPEDEFS
 #endif
@@ -464,7 +469,7 @@ typedef enum {
   CURLINFO_END
 } curl_infotype;
 
-typedef int (*curl_debug_callback)
+typedef int (CURL_FUNCC * curl_debug_callback)
        (CURL *handle,      /* the handle/transfer this concerns */
         curl_infotype type, /* what kind of data */
         char *data,        /* points to the data */
@@ -690,9 +695,9 @@ typedef enum {
 #endif /*!CURL_NO_OLDIES*/
 
 /* This prototype applies to all conversion callbacks */
-typedef CURLcode (*curl_conv_callback)(char *buffer, size_t length);
+typedef CURLcode (CURL_FUNCC * curl_conv_callback)(char *buffer, size_t length);
 
-typedef CURLcode (*curl_ssl_ctx_callback)(CURL *curl,    /* easy handle */
+typedef CURLcode (CURL_FUNCC * curl_ssl_ctx_callback)(CURL *curl,    /* easy handle */
                                           void *ssl_ctx, /* actually an OpenSSL
                                                             or WolfSSL SSL_CTX,
                                                             or an mbedTLS
@@ -801,7 +806,7 @@ enum curl_khmatch {
 };
 
 typedef int
-  (*curl_sshkeycallback) (CURL *easy,     /* easy handle */
+  (CURL_FUNCC *curl_sshkeycallback) (CURL *easy,     /* easy handle */
                           const struct curl_khkey *knownkey, /* known */
                           const struct curl_khkey *foundkey, /* found */
                           enum curl_khmatch, /* libcurl's view on the keys */
@@ -2101,8 +2106,8 @@ typedef enum {
 
 /* curl_strequal() and curl_strnequal() are subject for removal in a future
    release */
-CURL_EXTERN int curl_strequal(const char *s1, const char *s2);
-CURL_EXTERN int curl_strnequal(const char *s1, const char *s2, size_t n);
+CURL_EXTERN int (CURL_FUNCC curl_strequal)(const char *s1, const char *s2);
+CURL_EXTERN int (CURL_FUNCC curl_strnequal)(const char *s1, const char *s2, size_t n);
 
 /* Mime/form handling support. */
 typedef struct curl_mime_s      curl_mime;      /* Mime context. */
@@ -2303,7 +2308,7 @@ typedef enum {
  * adds one part that together construct a full post. Then use
  * CURLOPT_HTTPPOST to send it off to libcurl.
  */
-CURL_EXTERN CURLFORMcode curl_formadd(struct curl_httppost **httppost,
+CURL_EXTERN CURLFORMcode CURL_FUNCC curl_formadd(struct curl_httppost **httppost,
                                       struct curl_httppost **last_post,
                                       ...);
 
@@ -2315,8 +2320,7 @@ CURL_EXTERN CURLFORMcode curl_formadd(struct curl_httppost **httppost,
  * Should return the buffer length passed to it as the argument "len" on
  *   success.
  */
-typedef size_t (*curl_formget_callback)(void *arg, const char *buf,
-                                        size_t len);
+typedef size_t (CURL_FUNCC * curl_formget_callback)(void *arg, const char *buf, size_t len);
 
 /*
  * NAME curl_formget()
@@ -2328,7 +2332,7 @@ typedef size_t (*curl_formget_callback)(void *arg, const char *buf,
  * the curl_formget_callback function.
  * Returns 0 on success.
  */
-CURL_EXTERN int curl_formget(struct curl_httppost *form, void *arg,
+CURL_EXTERN int CURL_FUNCC curl_formget(struct curl_httppost *form, void *arg,
                              curl_formget_callback append);
 /*
  * NAME curl_formfree()
@@ -2337,7 +2341,7 @@ CURL_EXTERN int curl_formget(struct curl_httppost *form, void *arg,
  *
  * Free a multipart formpost previously built with curl_formadd().
  */
-CURL_EXTERN void curl_formfree(struct curl_httppost *form);
+CURL_EXTERN void CURL_FUNCC curl_formfree(struct curl_httppost *form);
 
 /*
  * NAME curl_getenv()
@@ -2347,7 +2351,7 @@ CURL_EXTERN void curl_formfree(struct curl_httppost *form);
  * Returns a malloc()'ed string that MUST be curl_free()ed after usage is
  * complete. DEPRECATED - see lib/README.curlx
  */
-CURL_EXTERN char *curl_getenv(const char *variable);
+CURL_EXTERN char * CURL_FUNCC curl_getenv(const char *variable);
 
 /*
  * NAME curl_version()
@@ -2356,7 +2360,7 @@ CURL_EXTERN char *curl_getenv(const char *variable);
  *
  * Returns a static ascii string of the libcurl version.
  */
-CURL_EXTERN char *curl_version(void);
+CURL_EXTERN char * CURL_FUNCC curl_version(void);
 
 /*
  * NAME curl_easy_escape()
@@ -2367,12 +2371,12 @@ CURL_EXTERN char *curl_version(void);
  * %XX versions). This function returns a new allocated string or NULL if an
  * error occurred.
  */
-CURL_EXTERN char *curl_easy_escape(CURL *handle,
+CURL_EXTERN char * CURL_FUNCC curl_easy_escape(CURL *handle,
                                    const char *string,
                                    int length);
 
 /* the previous version: */
-CURL_EXTERN char *curl_escape(const char *string,
+CURL_EXTERN char * CURL_FUNCC curl_escape(const char *string,
                               int length);
 
 
@@ -2387,13 +2391,13 @@ CURL_EXTERN char *curl_escape(const char *string,
  * Conversion Note: On non-ASCII platforms the ASCII %XX codes are
  * converted into the host encoding.
  */
-CURL_EXTERN char *curl_easy_unescape(CURL *handle,
+CURL_EXTERN char *CURL_FUNCC curl_easy_unescape(CURL *handle,
                                      const char *string,
                                      int length,
                                      int *outlength);
 
 /* the previous version */
-CURL_EXTERN char *curl_unescape(const char *string,
+CURL_EXTERN char * CURL_FUNCC curl_unescape(const char *string,
                                 int length);
 
 /*
@@ -2404,7 +2408,7 @@ CURL_EXTERN char *curl_unescape(const char *string,
  * Provided for de-allocation in the same translation unit that did the
  * allocation. Added in libcurl 7.10
  */
-CURL_EXTERN void curl_free(void *p);
+CURL_EXTERN void CURL_FUNCC curl_free(void *p);
 
 /*
  * NAME curl_global_init()
@@ -2416,7 +2420,7 @@ CURL_EXTERN void curl_free(void *p);
  *
  * This function is not thread-safe!
  */
-CURL_EXTERN CURLcode curl_global_init(long flags);
+CURL_EXTERN CURLcode CURL_FUNCC curl_global_init(long flags);
 
 /*
  * NAME curl_global_init_mem()
@@ -2431,7 +2435,7 @@ CURL_EXTERN CURLcode curl_global_init(long flags);
  * callback routines with be invoked by this library instead of the system
  * memory management routines like malloc, free etc.
  */
-CURL_EXTERN CURLcode curl_global_init_mem(long flags,
+CURL_EXTERN CURLcode CURL_FUNCC curl_global_init_mem(long flags,
                                           curl_malloc_callback m,
                                           curl_free_callback f,
                                           curl_realloc_callback r,
@@ -2446,7 +2450,7 @@ CURL_EXTERN CURLcode curl_global_init_mem(long flags,
  * curl_global_cleanup() should be invoked exactly once for each application
  * that uses libcurl
  */
-CURL_EXTERN void curl_global_cleanup(void);
+CURL_EXTERN void CURL_FUNCC curl_global_cleanup(void);
 
 /* linked-list structure for the CURLOPT_QUOTE option (and other) */
 struct curl_slist {
@@ -2503,7 +2507,7 @@ CURL_EXTERN CURLsslset curl_global_sslset(curl_sslbackend id, const char *name,
  * Appends a string to a linked list. If no list exists, it will be created
  * first. Returns the new list, after appending.
  */
-CURL_EXTERN struct curl_slist *curl_slist_append(struct curl_slist *,
+CURL_EXTERN struct curl_slist * CURL_FUNCC curl_slist_append(struct curl_slist *,
                                                  const char *);
 
 /*
@@ -2513,7 +2517,7 @@ CURL_EXTERN struct curl_slist *curl_slist_append(struct curl_slist *,
  *
  * free a previously built curl_slist.
  */
-CURL_EXTERN void curl_slist_free_all(struct curl_slist *);
+CURL_EXTERN void CURL_FUNCC curl_slist_free_all(struct curl_slist *);
 
 /*
  * NAME curl_getdate()
@@ -2524,7 +2528,7 @@ CURL_EXTERN void curl_slist_free_all(struct curl_slist *);
  * the first argument. The time argument in the second parameter is unused
  * and should be set to NULL.
  */
-CURL_EXTERN time_t curl_getdate(const char *p, const time_t *unused);
+CURL_EXTERN time_t CURL_FUNCC curl_getdate(const char *p, const time_t *unused);
 
 /* info about the certificate chain, only for OpenSSL, GnuTLS, Schannel, NSS
    and GSKit builds. Asked for with CURLOPT_CERTINFO / CURLINFO_CERTINFO */
@@ -2679,11 +2683,11 @@ typedef enum {
   CURL_LOCK_ACCESS_LAST        /* never use */
 } curl_lock_access;
 
-typedef void (*curl_lock_function)(CURL *handle,
+typedef void (CURL_FUNCC * curl_lock_function)(CURL *handle,
                                    curl_lock_data data,
                                    curl_lock_access locktype,
                                    void *userptr);
-typedef void (*curl_unlock_function)(CURL *handle,
+typedef void (CURL_FUNCC * curl_unlock_function)(CURL *handle,
                                      curl_lock_data data,
                                      void *userptr);
 
@@ -2709,9 +2713,9 @@ typedef enum {
   CURLSHOPT_LAST  /* never use */
 } CURLSHoption;
 
-CURL_EXTERN CURLSH *curl_share_init(void);
-CURL_EXTERN CURLSHcode curl_share_setopt(CURLSH *, CURLSHoption option, ...);
-CURL_EXTERN CURLSHcode curl_share_cleanup(CURLSH *);
+CURL_EXTERN CURLSH * CURL_FUNCC curl_share_init(void);
+CURL_EXTERN CURLSHcode CURL_FUNCC curl_share_setopt(CURLSH *, CURLSHoption option, ...);
+CURL_EXTERN CURLSHcode CURL_FUNCC curl_share_cleanup(CURLSH *);
 
 /****************************************************************************
  * Structures for querying information about the curl library at runtime.
@@ -2815,7 +2819,7 @@ typedef struct {
  * This function returns a pointer to a static copy of the version info
  * struct. See above.
  */
-CURL_EXTERN curl_version_info_data *curl_version_info(CURLversion);
+CURL_EXTERN curl_version_info_data * CURL_FUNCC curl_version_info(CURLversion);
 
 /*
  * NAME curl_easy_strerror()
@@ -2826,7 +2830,7 @@ CURL_EXTERN curl_version_info_data *curl_version_info(CURLversion);
  * into the equivalent human readable error string.  This is useful
  * for printing meaningful error messages.
  */
-CURL_EXTERN const char *curl_easy_strerror(CURLcode);
+CURL_EXTERN const char * CURL_FUNCC curl_easy_strerror(CURLcode);
 
 /*
  * NAME curl_share_strerror()
@@ -2837,7 +2841,7 @@ CURL_EXTERN const char *curl_easy_strerror(CURLcode);
  * into the equivalent human readable error string.  This is useful
  * for printing meaningful error messages.
  */
-CURL_EXTERN const char *curl_share_strerror(CURLSHcode);
+CURL_EXTERN const char * CURL_FUNCC curl_share_strerror(CURLSHcode);
 
 /*
  * NAME curl_easy_pause()
@@ -2848,7 +2852,7 @@ CURL_EXTERN const char *curl_share_strerror(CURLSHcode);
  * state by setting the bitmask, use the convenience defines below.
  *
  */
-CURL_EXTERN CURLcode curl_easy_pause(CURL *handle, int bitmask);
+CURL_EXTERN CURLcode CURL_FUNCC curl_easy_pause(CURL *handle, int bitmask);
 
 #define CURLPAUSE_RECV      (1<<0)
 #define CURLPAUSE_RECV_CONT (0)
diff --git a/include/curl/easy.h b/include/curl/easy.h
index 592f5d3..e951bd9 100644
--- a/include/curl/easy.h
+++ b/include/curl/easy.h
@@ -25,10 +25,10 @@
 extern "C" {
 #endif
 
-CURL_EXTERN CURL *curl_easy_init(void);
-CURL_EXTERN CURLcode curl_easy_setopt(CURL *curl, CURLoption option, ...);
-CURL_EXTERN CURLcode curl_easy_perform(CURL *curl);
-CURL_EXTERN void curl_easy_cleanup(CURL *curl);
+CURL_EXTERN CURL * CURL_FUNCC curl_easy_init(void);
+CURL_EXTERN CURLcode CURL_FUNCC curl_easy_setopt(CURL *curl, CURLoption option, ...);
+CURL_EXTERN CURLcode CURL_FUNCC curl_easy_perform(CURL *curl);
+CURL_EXTERN void CURL_FUNCC curl_easy_cleanup(CURL *curl);
 
 /*
  * NAME curl_easy_getinfo()
@@ -43,7 +43,7 @@ CURL_EXTERN void curl_easy_cleanup(CURL *curl);
  * performed transfer, all results from this function are undefined until the
  * transfer is completed.
  */
-CURL_EXTERN CURLcode curl_easy_getinfo(CURL *curl, CURLINFO info, ...);
+CURL_EXTERN CURLcode CURL_FUNCC curl_easy_getinfo(CURL *curl, CURLINFO info, ...);
 
 
 /*
@@ -58,7 +58,7 @@ CURL_EXTERN CURLcode curl_easy_getinfo(CURL *curl, CURLINFO info, ...);
  * curl_easy_duphandle() for each new thread to avoid a series of identical
  * curl_easy_setopt() invokes in every thread.
  */
-CURL_EXTERN CURL *curl_easy_duphandle(CURL *curl);
+CURL_EXTERN CURL *CURL_FUNCC curl_easy_duphandle(CURL *curl);
 
 /*
  * NAME curl_easy_reset()
@@ -71,7 +71,7 @@ CURL_EXTERN CURL *curl_easy_duphandle(CURL *curl);
  * It does keep: live connections, the Session ID cache, the DNS cache and the
  * cookies.
  */
-CURL_EXTERN void curl_easy_reset(CURL *curl);
+CURL_EXTERN void CURL_FUNCC curl_easy_reset(CURL *curl);
 
 /*
  * NAME curl_easy_recv()
@@ -81,7 +81,7 @@ CURL_EXTERN void curl_easy_reset(CURL *curl);
  * Receives data from the connected socket. Use after successful
  * curl_easy_perform() with CURLOPT_CONNECT_ONLY option.
  */
-CURL_EXTERN CURLcode curl_easy_recv(CURL *curl, void *buffer, size_t buflen,
+CURL_EXTERN CURLcode CURL_FUNCC curl_easy_recv(CURL *curl, void *buffer, size_t buflen,
                                     size_t *n);
 
 /*
diff --git a/include/curl/multi.h b/include/curl/multi.h
index bda9bb7..8bf53d6 100644
--- a/include/curl/multi.h
+++ b/include/curl/multi.h
@@ -124,7 +124,7 @@ struct curl_waitfd {
  *
  * Returns: a new CURLM handle to use in all 'curl_multi' functions.
  */
-CURL_EXTERN CURLM *curl_multi_init(void);
+CURL_EXTERN CURLM *CURL_FUNCC curl_multi_init(void);
 
 /*
  * Name:    curl_multi_add_handle()
@@ -133,7 +133,7 @@ CURL_EXTERN CURLM *curl_multi_init(void);
  *
  * Returns: CURLMcode type, general multi error code.
  */
-CURL_EXTERN CURLMcode curl_multi_add_handle(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_add_handle(CURLM *multi_handle,
                                             CURL *curl_handle);
 
  /*
@@ -143,7 +143,7 @@ CURL_EXTERN CURLMcode curl_multi_add_handle(CURLM *multi_handle,
   *
   * Returns: CURLMcode type, general multi error code.
   */
-CURL_EXTERN CURLMcode curl_multi_remove_handle(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_remove_handle(CURLM *multi_handle,
                                                CURL *curl_handle);
 
  /*
@@ -155,7 +155,7 @@ CURL_EXTERN CURLMcode curl_multi_remove_handle(CURLM *multi_handle,
   *
   * Returns: CURLMcode type, general multi error code.
   */
-CURL_EXTERN CURLMcode curl_multi_fdset(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_fdset(CURLM *multi_handle,
                                        fd_set *read_fd_set,
                                        fd_set *write_fd_set,
                                        fd_set *exc_fd_set,
@@ -169,7 +169,7 @@ CURL_EXTERN CURLMcode curl_multi_fdset(CURLM *multi_handle,
  *
  * Returns:  CURLMcode type, general multi error code.
  */
-CURL_EXTERN CURLMcode curl_multi_wait(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_wait(CURLM *multi_handle,
                                       struct curl_waitfd extra_fds[],
                                       unsigned int extra_nfds,
                                       int timeout_ms,
@@ -214,7 +214,7 @@ CURL_EXTERN CURLMcode curl_multi_wakeup(CURLM *multi_handle);
   *          still have occurred problems on individual transfers even when
   *          this returns OK.
   */
-CURL_EXTERN CURLMcode curl_multi_perform(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_perform(CURLM *multi_handle,
                                          int *running_handles);
 
  /*
@@ -227,7 +227,7 @@ CURL_EXTERN CURLMcode curl_multi_perform(CURLM *multi_handle,
   *
   * Returns: CURLMcode type, general multi error code.
   */
-CURL_EXTERN CURLMcode curl_multi_cleanup(CURLM *multi_handle);
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_cleanup(CURLM *multi_handle);
 
 /*
  * Name:    curl_multi_info_read()
@@ -257,7 +257,7 @@ CURL_EXTERN CURLMcode curl_multi_cleanup(CURLM *multi_handle);
  *          queue (after this read) in the integer the second argument points
  *          to.
  */
-CURL_EXTERN CURLMsg *curl_multi_info_read(CURLM *multi_handle,
+CURL_EXTERN CURLMsg *CURL_FUNCC curl_multi_info_read(CURLM *multi_handle,
                                           int *msgs_in_queue);
 
 /*
@@ -269,7 +269,7 @@ CURL_EXTERN CURLMsg *curl_multi_info_read(CURLM *multi_handle,
  *
  * Returns: A pointer to a zero-terminated error message.
  */
-CURL_EXTERN const char *curl_multi_strerror(CURLMcode);
+CURL_EXTERN const char *CURL_FUNCC curl_multi_strerror(CURLMcode);
 
 /*
  * Name:    curl_multi_socket() and
@@ -292,7 +292,7 @@ CURL_EXTERN const char *curl_multi_strerror(CURLMcode);
 #define CURL_CSELECT_OUT  0x02
 #define CURL_CSELECT_ERR  0x04
 
-typedef int (*curl_socket_callback)(CURL *easy,      /* easy handle */
+typedef int (CURL_FUNCC *curl_socket_callback)(CURL *easy,      /* easy handle */
                                     curl_socket_t s, /* socket */
                                     int what,        /* see above */
                                     void *userp,     /* private callback
@@ -309,20 +309,20 @@ typedef int (*curl_socket_callback)(CURL *easy,      /* easy handle */
  *
  * Returns: The callback should return zero.
  */
-typedef int (*curl_multi_timer_callback)(CURLM *multi,    /* multi handle */
+typedef int (CURL_FUNCC *curl_multi_timer_callback)(CURLM *multi,    /* multi handle */
                                          long timeout_ms, /* see above */
                                          void *userp);    /* private callback
                                                              pointer */
 
-CURL_EXTERN CURLMcode curl_multi_socket(CURLM *multi_handle, curl_socket_t s,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_socket(CURLM *multi_handle, curl_socket_t s,
                                         int *running_handles);
 
-CURL_EXTERN CURLMcode curl_multi_socket_action(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_socket_action(CURLM *multi_handle,
                                                curl_socket_t s,
                                                int ev_bitmask,
                                                int *running_handles);
 
-CURL_EXTERN CURLMcode curl_multi_socket_all(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_socket_all(CURLM *multi_handle,
                                             int *running_handles);
 
 #ifndef CURL_ALLOW_OLD_MULTI_SOCKET
@@ -408,7 +408,7 @@ typedef enum {
  *
  * Returns: CURLM error code.
  */
-CURL_EXTERN CURLMcode curl_multi_setopt(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_setopt(CURLM *multi_handle,
                                         CURLMoption option, ...);
 
 
@@ -421,7 +421,7 @@ CURL_EXTERN CURLMcode curl_multi_setopt(CURLM *multi_handle,
  *
  * Returns: CURLM error code.
  */
-CURL_EXTERN CURLMcode curl_multi_assign(CURLM *multi_handle,
+CURL_EXTERN CURLMcode CURL_FUNCC curl_multi_assign(CURLM *multi_handle,
                                         curl_socket_t sockfd, void *sockp);
 
 
-- 
2.20.1

